File: BukkitUtil-core/src/main/java/no/kh498/util/jackson/serializers/bean/ConfigurationSerializableSerializer.java
package no.kh498.util.jackson.serializers.bean;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;
import org.bukkit.configuration.serialization.ConfigurationSerializable;

import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.bukkit.configuration.serialization.ConfigurationSerialization.SERIALIZED_TYPE_KEY;
import static org.bukkit.configuration.serialization.ConfigurationSerialization.getAlias;

/**
 * @author Elg
 */
public class ConfigurationSerializableSerializer extends StdSerializer<ConfigurationSerializable> {

    static { inst = new ConfigurationSerializableSerializer(); }

    public static final ConfigurationSerializableSerializer inst;

    protected ConfigurationSerializableSerializer() {
        super(ConfigurationSerializable.class);
    }

    @Override
    public void serialize(ConfigurationSerializable ser, JsonGenerator gen, SerializerProvider provider)
    throws IOException {


        Map<String, Object> map = new LinkedHashMap<>();
        map.put(SERIALIZED_TYPE_KEY, getAlias(ser.getClass()));
        map.putAll(ser.serialize());
        gen.writeObject(map);
    }

//    private Object serializeConfSer(Object ser) {
//
//    }
//
//    private class ConfigurationSerializableWrapper implements ConfigurationSerializable {
//
//        private final Object o;
//
//        ConfigurationSerializableWrapper(Object o) {this.o = o;}
//
//        @Override
//        public Map<String, Object> serialize() {
//            Map<String, Object> map = new HashMap<>();
//            YamlConfiguration conf = new YamlConfiguration();
//            conf. return map;
//        }
//    }
}

File: BukkitUtil-core/src/main/kotlin/no/kh498/util/jackson/serializers/bean/ConfigurationSerializableSerializer.kt
package no.kh498.util.jackson.serializers.bean

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.TSFBuilder
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.jsontype.TypeSerializer
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLFactoryBuilder
import com.fasterxml.jackson.dataformat.yaml.YAMLParser
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.configuration.file.YamlConstructor
import org.bukkit.configuration.file.YamlRepresenter
import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.yaml.snakeyaml.DumperOptions
import org.yaml.snakeyaml.Yaml
import java.io.IOException

/**
 * @author Elg
 */
object ConfigurationSerializableSerializer : StdSerializer<ConfigurationSerializable>(ConfigurationSerializable::class.java) {

    const val ROOT_PATH = "root"

    @Throws(IOException::class)
    override fun serialize(ser: ConfigurationSerializable, gen: JsonGenerator, provider: SerializerProvider) {
        gen.writeObject(ser.serialize())
    }

    override fun serializeWithType(value: ConfigurationSerializable?, gen: JsonGenerator?, serializers: SerializerProvider?, typeSer: TypeSerializer?) {
        super.serializeWithType(value, gen, serializers, typeSer)
    }
}
